mantle/cmd/kola/devshell.go

```diff
@@ -0,0 +1,303 @@
+		return errors.Wrapf(err, "rendering config")
+	}
+
+	builder.InheritConsole = devshellConsole
+	inst, err := builder.Exec()
+	if err != nil {
+		return err
+	}
+	defer inst.Destroy()
+
+	if devshellConsole {
+		return inst.Wait()
+	}
+
+	qemuWaitChan := make(chan error)
+	errchan := make(chan error)
+	readychan := make(chan struct{})
+	go func() {
+		buf, err := inst.WaitIgnitionError()
+		if err != nil {
+			errchan <- err
+		} else {
+			// TODO parse buf and try to nicely render something
+			if buf != "" {
+				errchan <- platform.ErrInitramfsEmergency
+			}
+		}
+	}()
+	go func() {
+		qemuWaitChan <- inst.Wait()
+	}()
+	go func() {
+		readyMsg, err := readTrimmedLine(readyReader)
+		if err != nil {
+			errchan <- err
+		}
+		if readyMsg != "ready" {
+			errchan <- fmt.Errorf("Unexpected ready message: %s", readyMsg)
+		}
+		var s struct{}
+		readychan <- s
+	}()
+
+	select {
+	case err := <-errchan:
+		if err == platform.ErrInitramfsEmergency {
+			return fmt.Errorf("instance failed in initramfs; try rerunning with --devshell-console")
+		}
+		return err
+	case err := <-qemuWaitChan:
+		return errors.Wrapf(err, "qemu exited before setup")
+	case _ = <-readychan:
+		fmt.Println("virtio: connected")
+	}
+
+	var ip string
+	err = util.Retry(6, 5*time.Second, func() error {
+		var err error
+		ip, err = inst.SSHAddress()
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "awaiting ssh address")
+	}
+
+	sshConfigPath := filepath.Join(tmpd, "ssh-config")
+	sshConfig, err := os.OpenFile(sshConfigPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return errors.Wrapf(err, "creating ssh config")
+	}
+	defer sshConfig.Close()
+	sshBuf := bufio.NewWriter(sshConfig)
+
+	_, err = fmt.Fprintf(sshBuf, "Host %s\n", devshellHostname)
+	if err != nil {
+		return err
+	}
+	host, port, err := net.SplitHostPort(ip)
+	if err != nil {
+		// Yeah this is hacky, surprising there's not a stdlib API for this
+		host = ip
+		port = ""
+	}
+	if port != "" {
+		if _, err := fmt.Fprintf(sshBuf, "  Port %s\n", port); err != nil {
+			return err
+		}
+	}
+	if _, err := fmt.Fprintf(sshBuf, `  HostName %s
+	  StrictHostKeyChecking no
+	  UserKnownHostsFile /dev/null
+	  User core
+	  PasswordAuthentication no
+	  KbdInteractiveAuthentication no
+	  GSSAPIAuthentication no
+	  IdentitiesOnly yes
+	  ForwardAgent no
+	  ForwardX11 no
+	`, host); err != nil {
+		return err
+	}
+
+	if err := sshBuf.Flush(); err != nil {
+		return err
+	}
+
+	err = util.Retry(10, 1*time.Second, func() error {
+		cmd := devshellSSH(sshConfigPath, sshKeyPath, true, "true")
+		return cmd.Run()
+	})
+	if err != nil {
+		return err
+	}
+
+	poweroffStarted := false
+	go func() {
+		msg, _ := readTrimmedLine(readyReader)
+		if msg == "poweroff" {
+			poweroffStarted = true
+		}
+	}()
+
+	go func() {
+		for {
+			// FIXME figure out how to differentiate between reboot/shutdown
```

Comment:
Is this fixed already?  The `poweroff` / reboot functionality seemed to work for me.  `poweroff` terminated my session in both devshell and devshell-console.  `reboot` reconnected for devshell and stayed connected for devshell-console

Reply:
Yeah, that all works.  But what I tried to do is have a check when qemu terminates that we got the "expected shutdown" message.  It's not a super important FIXME.